Assets/Mirror/Runtime/Transport/LatencySimulation.cs

```diff
@@ -0,0 +1,274 @@
+    {
+        public Transport wrap;
+
+        [Header("Reliable Messages")]
+        [Tooltip("Reliable latency in seconds")]
+        public float reliableLatency = 0;
+        // note: packet loss over reliable manifests itself in latency.
+        //       don't need (and can't add) a loss option here.
+        // note: reliable is ordered by definition. no need to scramble.
+
+        [Header("Unreliable Messages")]
+        [Tooltip("Packet loss in %")]
+        [Range(0, 1)] public float unreliableLoss;
+        [Tooltip("Unreliable latency in seconds")]
+        public float unreliableLatency = 0;
+        [Tooltip("Scramble unreliable messages, just like over the real network. Mirror unreliable is unordered.")]
+        public bool unreliableScramble;
+
+        // message queues
+        // list so we can insert randomly (scramble)
+        List<QueuedMessage> reliableClientToServer = new List<QueuedMessage>();
+        List<QueuedMessage> reliableServerToClient = new List<QueuedMessage>();
+        List<QueuedMessage> unreliableClientToServer = new List<QueuedMessage>();
+        List<QueuedMessage> unreliableServerToClient = new List<QueuedMessage>();
+
+        // random
+        // UnityEngine.Random.value is [0, 1] but we need [0, 1)
+        // aka exclusive to 1, not inclusive.
+        // => NextDouble() is NEVER < 0 so loss=0 never drops!
+        // => NextDouble() is ALWAYS < 1 so loss=1 always drops!
+        System.Random random = new System.Random();
+
+        public void Awake()
+        {
+            if (wrap == null)
+                throw new Exception("PressureDrop requires an underlying transport to wrap around.");
+        }
+
+        // forward enable/disable to the wrapped transport
+        void OnEnable() { wrap.enabled = true; }
+        void OnDisable() { wrap.enabled = false; }
+
+        public override bool Available() => wrap.Available();
+
+        public override void ClientConnect(string address)
+        {
+            wrap.OnClientConnected = OnClientConnected;
+            wrap.OnClientDataReceived = OnClientDataReceived;
+            wrap.OnClientError = OnClientError;
+            wrap.OnClientDisconnected = OnClientDisconnected;
+            wrap.ClientConnect(address);
+        }
+
+        public override void ClientConnect(Uri uri)
+        {
+            wrap.OnClientConnected = OnClientConnected;
+            wrap.OnClientDataReceived = OnClientDataReceived;
+            wrap.OnClientError = OnClientError;
+            wrap.OnClientDisconnected = OnClientDisconnected;
+            wrap.ClientConnect(uri);
+        }
+
+        public override bool ClientConnected() => wrap.ClientConnected();
+
+        public override void ClientDisconnect()
+        {
+            wrap.ClientDisconnect();
+            reliableClientToServer.Clear();
+            unreliableClientToServer.Clear();
+        }
+
+        public override void ClientSend(int channelId, ArraySegment<byte> segment)
```

Comment:
ClientSend and ServerSend appear to be identical except for 1 line.
Perhaps create a single method they both call with a server bool to determine whether to do
`unreliableClientToServer.Insert(index, message);`
or 
`unreliableServerToClient.Insert(index, message);`

Reply:
yep, nice idea. helper function added.